### ISSUE TITLE
Android -- No OpKernel was registered to support Op 'SparseToDense' with these attrs

### ISSUE THREAD

**Issue by selcouthlyBlue** [tensorflow/tensorflow#13651]
I am trying to load a graph inside Android that I generated and frozen. I keep getting this error whenever I try to run it:

```
Caused by: java.lang.IllegalArgumentException: No OpKernel was registered to support
Op 'SparseToDense' with these attrs.  Registered devices: [CPU], Registered kernels:
device='CPU'; T in [DT_STRING]; Tindices in [DT_INT64]
device='CPU'; T in [DT_STRING]; Tindices in [DT_INT32]
device='CPU'; T in [DT_BOOL]; Tindices in [DT_INT64]
device='CPU'; T in [DT_BOOL]; Tindices in [DT_INT32]
device='CPU'; T in [DT_FLOAT]; Tindices in [DT_INT64]
device='CPU'; T in [DT_FLOAT]; Tindices in [DT_INT32]
device='CPU'; T in [DT_INT32]; Tindices in [DT_INT64]
device='CPU'; T in [DT_INT32]; Tindices in [DT_INT32]
                                                                                            
[[Node: output = SparseToDense[T=DT_INT64, Tindices=DT_INT64, validate_indices=true](CTCBeamSearchDecoder, CTCBeamSearchDecoder:2, CTCBeamSearchDecoder:1, output/default_value)]]
```

I also optimized the graph for inference. When I try to load it onto Android, I get this error:

```
java.io.IOException: Not a valid TensorFlow Graph serialization: NodeDef expected inputs '' do not match 1 inputs 
specified; Op<name=Const; signature= -> output:dtype; attr=value:tensor; attr=dtype:type>; 
NodeDef: stack_bidirectional_rnn/cell_0/bidirectional_rnn/bw/bw/while/add/y = Const[dtype=DT_INT32, 
value=Tensor<type: int32 shape: [] values: 1>](stack_bidirectional_rnn/cell_0/bidirectional_rnn/bw/bw/while/Switch:1)
```

Steps to get the frozen graph and optimized graph:
1. Clone this [repository](https://github.com/selcouthlyBlue/bi_lstm_ocr):
2. Run dummy_train.py producing the .pbtxt and checkpoint files
3. Run dummy_freeze_and_save.py producing the frozen and optimized graphs frozen_bi_lstm_ctc_ocr.pb and optimized_frozen_bi_lstm_ctc_ocr.pb, respectively.

Files in the mentioned repository relevant to the problem:
- TFNetwork.py (contains the network specs and training code)
- utils.py (containing the graph freezing and optimization codes)
- The dummy configs in the test files (just to train the model for 1 epoch)

Here is the Java part related to the problem:

   ```
 public String recognizeHandwritingFrom(Bitmap bitmap) {
        bitmap = Bitmap.createScaledBitmap(bitmap, 1024, 128,, true);
        int[] intValues = new int[bitmap.getWidth() * bitmap.getHeight()];
        bitmap.getPixels(intValues, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(), bitmap.getHeight());
        float[] floatValues = new float[bitmap.getWidth() * bitmap.getHeight()];
        for (int i = 0; i < intValues.length; ++i) {
            final int val = intValues[i];
            floatValues[i] = (((val >> 16) & 0xFF));
        }
        float[] result = new float[80];

        long[] INPUT_SIZE = new long[]{1, bitmap.getHeight(), bitmap.getWidth()};
        String[] inputs = new String[]{"input", "seq_len_input"};
        inferenceInterface.feed(inputs[0], floatValues, INPUT_SIZE);
        inferenceInterface.feed(inputs[1], new int[]{bitmap.getWidth()}, 1);

        String[] outputs = new String[]{"output"};
        inferenceInterface.run(outputs);
        inferenceInterface.fetch(outputs[0], result);

        return result.toString();
    }
```
I'm using 

- Python 3.5.3 :: Anaconda custom (64-bit)
- The python tensorflow build is downloaded using Anaconda. Tensorflow version is 1.2.1
- The compiled tensorflow for IOS is from this [nightly build](https://ci.tensorflow.org/view/Nightly/job/nightly-android/44/artifact/).

I would be really grateful if anyone has an idea on why Android seems to not be able to find `SparseToDense` as this is the only thing I have to fix to make it work.

If you would like to run the android application as well, you can clone this [repository](https://github.com/selcouthlyBlue/mem2speech). Just get the files from the nightly build and place them inside app/libs folder following this structure:

```
libs
|____arm64-v8a
| |____libtensorflow_inference.so
|____armeabi-v7a
| |____libtensorflow_inference.so
|____libandroid_tensorflow_inference_java.jar
|____x86
| |____libtensorflow_inference.so
|____x86_64
| |____libtensorflow_inference.so
```

**Comment by reedwm** [comment]
@petewarden can you take a look?

**Comment by selcouthlyBlue** [comment]
Thank you for bringing attention to this :D

**Comment by eli99999** [comment]
I'm encountering the same problem.
I see that "sparse_to_dense_op" is in the android_extended_ops_group2 in the core/kernel BUILD file, hence it is supposed to working in android

**Comment by eli99999** [comment]
encountered the same problem.
you are sending int64 where only int32 is accepted
solved it by adding casting to the python code that is building the graph
decoded, _ = tf.nn.ctc_beam_search_decoder(logit.sg_transpose(perm=[1, 0, 2]), seq_len, merge_repeated=False)
	y = tf.sparse_to_dense( tf.to_int32(decoded[0].indices),  tf.to_int32(decoded[0].dense_shape),  tf.to_int32(decoded[0].values), name="output_node" )

**Comment by petewarden** [comment]
I don't have an answer on the second optimization question unfortunately, but if you do want to support more than the standard float and int32 types on mobile, you can recompile TensorFlow with __ANDROID_TYPES_FULL__ defined. This will result in a larger binary though. We're working on improving the documentation around this, and there's some information in this free ebook too: http://www.oreilly.com/data/free/building-mobile-applications-with-tensorflow.csp

**Comment by selcouthlyBlue** [comment]
I apologize for the late reply. It's weird that gmail didn't update me of your replies. I tried @eli99999 's solution and now I got this error when I ran the application:

```
java.lang.IllegalArgumentException: Matrix size-incompatible: In[0]: [1,1056], In[1]: [160,128]
                                                 	 [[Node: stack_bidirectional_rnn/cell_0/bidirectional_rnn/bw/bw/while/bw/basic_lstm_cell/basic_lstm_cell/MatMul = MatMul[T=DT_FLOAT, transpose_a=false, transpose_b=false, _device="/job:localhost/replica:0/task:0/cpu:0"](stack_bidirectional_rnn/cell_0/bidirectional_rnn/bw/bw/while/bw/basic_lstm_cell/basic_lstm_cell/concat, stack_bidirectional_rnn/cell_0/bidirectional_rnn/bw/bw/while/bw/basic_lstm_cell/basic_lstm_cell/MatMul/Enter)]]
```

**Comment by selcouthlyBlue** [comment]
I did a little poking with the network parameters and it turns out, the number of hidden units influence the dimensions found in "In[1]". With that, here's the code building the bidirectional lstm part of the graph:

`
#num_layers = 3,
#num_hidden = 32
#num_classes = 80 

lstm_fw_cells = [tf.contrib.rnn.BasicLSTMCell(num_hidden, forget_bias=1.0) for _ in range(num_layers)]
lstm_bw_cells = [tf.contrib.rnn.BasicLSTMCell(num_hidden, forget_bias=1.0) for _ in range(num_layers)]

outputs, _, _ = tf.contrib.rnn.stack_bidirectional_dynamic_rnn(lstm_fw_cells, lstm_bw_cells,
                                                                   self.inputs, dtype=tf.float32)

batch_size = tf.shape(self.inputs)[0]

outputs = tf.reshape(outputs, [-1, num_hidden])

W = tf.Variable(tf.truncated_normal([num_hidden, num_classes], stddev=0.1, dtype=tf.float32))
b = tf.Variable(tf.constant(0., shape=[num_classes], dtype=tf.float32))

logits = tf.matmul(outputs, W) + b
logits = tf.reshape(logits, [batch_size, -1, num_classes])
logits = tf.transpose(logits, (1, 0, 2))`

**Comment by selcouthlyBlue** [comment]
I managed to fix the error from applying @eli99999 's solution by changing the bitmap image_width to 128 in the application's recognizer config. I don't know why but it just fixed that error.

```
Recognizer recognizer = new OfflineRecognizer(recognizerConfigBuilder.setAssetManager(context.getAssets())
                .setImageHeight(128)
                .setImageWidth(128)
                .setModelFilename("frozen_bi_lstm_ctc_ocr.pb")
                .setCharsetFile("chars.txt")
                .build());
```

With that change, I now encounter this new error:

`java.lang.IllegalArgumentException: cannot use java.nio.FloatArrayBuffer with Tensor of type INT32`

I'm guessing this is the part where @petewarden 's solution will work. @eli99999 is there a way for int64 to be accepted instead of int32?

**Comment by selcouthlyBlue** [comment]
I already fixed the error previous error. I just changed the type of result from float to int. But I have some issues with changing the image width to 128 just to get it working. For now, I'm going to close this issue since the error has been addressed.

**Comment by ramrahu** [comment]
> encountered the same problem.
> you are sending int64 where only int32 is accepted
> solved it by adding casting to the python code that is building the graph
> decoded, _ = tf.nn.ctc_beam_search_decoder(logit.sg_transpose(perm=[1, 0, 2]), seq_len, merge_repeated=False)
> y = tf.sparse_to_dense( tf.to_int32(decoded[0].indices), tf.to_int32(decoded[0].dense_shape), tf.to_int32(decoded[0].values), name="output_node" )

Hi,
I had the same issue and I tried this method. But after this, I started getting the following error:
2019-04-08 11:54:59.302 17408-17563/org.tensorflow.demo E/TensorFlowInferenceInterface: Failed to run TensorFlow inference with inputs:[SeqLen], outputs:[SparseToDense]
2019-04-08 11:54:59.302 17408-17563/org.tensorflow.demo E/AndroidRuntime: FATAL EXCEPTION: Thread-3
    Process: org.tensorflow.demo, PID: 17408
    java.lang.IllegalArgumentException: Expects arg[0] to be int32 but float is provided
        at org.tensorflow.Session.run(Native Method)
        at org.tensorflow.Session.access$100(Session.java:48)
        at org.tensorflow.Session$Runner.runHelper(Session.java:314)
        at org.tensorflow.Session$Runner.run(Session.java:264)
        at org.tensorflow.contrib.android.TensorFlowInferenceInterface.run(TensorFlowInferenceInterface.java:228)
        at org.tensorflow.contrib.android.TensorFlowInferenceInterface.run(TensorFlowInferenceInterface.java:197)
        at org.tensorflow.contrib.android.TensorFlowInferenceInterface.run(TensorFlowInferenceInterface.java:187)
        at org.tensorflow.demo.SpeechActivity.recognize(SpeechActivity.java:229)
        at org.tensorflow.demo.SpeechActivity.access$100(SpeechActivity.java:48)
        at org.tensorflow.demo.SpeechActivity$3.run(SpeechActivity.java:193)
        at java.lang.Thread.run(Thread.java:761)
Could you help me with this?
